server/src/main/java/org/elasticsearch/index/query/CombinedFieldsQueryBuilder.java

```diff
@@ -0,0 +1,433 @@
+            throw new IllegalArgumentException("["+ NAME + "] queries can only be used with the [BM25] similarity");
+        }
+    }
+
+    private static final class FieldAndBoost {
+        final MappedFieldType fieldType;
+        final float boost;
+
+        FieldAndBoost(MappedFieldType fieldType, float boost) {
+            this.fieldType = Objects.requireNonNull(fieldType);
+            this.boost = boost;
+        }
+    }
+
+    private static class CombinedFieldsBuilder extends QueryBuilder {
+        private final List<FieldAndBoost> fields;
+
+        CombinedFieldsBuilder(List<FieldAndBoost> fields,
+                              Analyzer analyzer,
+                              boolean autoGenerateSynonymsPhraseQuery) {
+            super(analyzer);
+            this.fields = fields;
+            setAutoGenerateMultiTermSynonymsPhraseQuery(autoGenerateSynonymsPhraseQuery);
+        }
+
+        @Override
+        protected Query createFieldQuery(TokenStream source, BooleanClause.Occur operator, String field, boolean quoted, int phraseSlop) {
+            if (source.hasAttribute(DisableGraphAttribute.class)) {
+                /*
+                 * A {@link TokenFilter} in this {@link TokenStream} disabled the graph analysis to avoid
+                 * paths explosion. See {@link org.elasticsearch.index.analysis.ShingleTokenFilterFactory} for details.
+                 */
+                setEnableGraphQueries(false);
+            }
+            try {
+                return super.createFieldQuery(source, operator, field, quoted, phraseSlop);
+            } finally {
+                setEnableGraphQueries(true);
+            }
+        }
+
+        @Override
+        public Query createPhraseQuery(String field, String queryText, int phraseSlop) {
+            throw new IllegalArgumentException("[combined_fields] queries don't support phrases");
+        }
+
+        @Override
+        protected Query newSynonymQuery(TermAndBoost[] terms) {
+            XCombinedFieldQuery.Builder query = new XCombinedFieldQuery.Builder();
+            for (TermAndBoost termAndBoost : terms) {
+                assert termAndBoost.boost == BoostAttribute.DEFAULT_BOOST;
+                BytesRef bytes = termAndBoost.term.bytes();
+                query.addTerm(bytes);
+            }
+            for (FieldAndBoost fieldAndBoost : fields) {
+                MappedFieldType fieldType = fieldAndBoost.fieldType;
+                float fieldBoost = fieldAndBoost.boost;
+                query.addField(fieldType.name(), fieldBoost);
+            }
+            return query.build();
+        }
+
+        @Override
+        protected Query newTermQuery(Term term, float boost) {
+            TermAndBoost termAndBoost = new TermAndBoost(term, boost);
+            return newSynonymQuery(new TermAndBoost[]{termAndBoost});
+        }
+
+        @Override
+        protected Query analyzePhrase(String field, TokenStream stream, int slop) throws IOException {
```

Comment:
Is this method ever called if we don't support phrases in a `combined_fields` query?